src/celeritas/neutron/interactor/NeutronElasticInteractor.hh

```diff
@@ -0,0 +1,155 @@
+    NeutronElasticInteractor(NeutronElasticRef const& shared,
+                             ParticleTrackView const& particle,
+                             Real3 const& inc_direction,
+                             IsotopeView const& target);
+
+    // Sample an interaction with the given RNG
+    template<class Engine>
+    inline CELER_FUNCTION Interaction operator()(Engine& rng);
+
+  private:
+    //// TYPES ////
+
+    using UniformRealDist = UniformRealDistribution<real_type>;
+
+    //// DATA ////
+
+    // Constant shared data
+    NeutronElasticRef const& shared_;
+    // Incident neutron energy
+    units::MevEnergy const inc_energy_;
+    // Incident neutron direction
+    Real3 const& inc_direction_;
+    // Target nucleus
+    IsotopeView const& target_;
+
+    // Value of neutron MeVMass
+    real_type neutron_mass_;
+    real_type neutron_p_;
+
+    // Sampler
+    UniformRealDist sample_phi_;
+    detail::InvariantQ2Sampler sample_Q2_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with shared and state data, and a target nucleus.
+ */
+CELER_FUNCTION NeutronElasticInteractor::NeutronElasticInteractor(
+    NeutronElasticRef const& shared,
+    ParticleTrackView const& particle,
+    Real3 const& inc_direction,
+    IsotopeView const& target)
+    : shared_(shared)
+    , inc_energy_(particle.energy())
+    , inc_direction_(inc_direction)
+    , target_(target)
+    , neutron_mass_(value_as<Mass>(shared_.neutron_mass))
+    , neutron_p_(std::sqrt(inc_energy_.value()
+                           * (inc_energy_.value() + 2 * neutron_mass_)))
+    , sample_phi_(0, 2 * constants::pi)
+    , sample_Q2_(shared_, target_, neutron_p_)
+{
+    CELER_EXPECT(particle.particle_id() == shared_.ids.neutron);
+    CELER_EXPECT(inc_energy_ > zero_quantity());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the final state of the neutron-nucleus elastic scattering.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction NeutronElasticInteractor::operator()(Engine& rng)
+{
+    using Energy = units::MevEnergy;
+
+    // Scattered neutron with respect to the axis of incident direction
+    Interaction result;
+
+    // The momentum magnitude in the c.m. frame
+    real_type target_mass = value_as<Mass>(target_.nuclear_mass());
+    real_type neutron_energy = neutron_mass_ + inc_energy_.value();
+    real_type cm_p = neutron_p_
+                     / std::sqrt(1 + ipow<2>(neutron_mass_ / target_mass)
+                                 + 2 * neutron_energy / target_mass);
+
+    // Sample the scattered direction from the invariant momentum transfer
+    // squared (\f$ -t = Q^{2} \f$) in the c.m. frame
+    real_type cos_theta = 1 - real_type(0.5) * sample_Q2_(rng) / ipow<2>(cm_p);
+    clamp(cos_theta, real_type{-1}, real_type{1});
```

Comment:
The `clamp` operation is "pure" , i.e. it doesn't modify `cos_theta` but returns a clamped result, so this is a null-op. I think it should also be the burden of `sample_q2` to `CELER_ENSURE` that the result is in `[0, 2 * cm_p^2]`, and this should be an assertion here rather than a mandatory clamp.